feat: expose internal TrayIcon

petersamokhin opened the pull request:
Hello there! Thanks a lot for Tauri.

This PR exposes the internal TrayIcon to be able to access the platform implementations.
This is pretty important and my use case is to access the underlying `ns_status_item()` (and then NSWindow, and then NSScreen). TrayIcon already exposes that, but it's unavailable if using the Tauri plugin.

petersamokhin committed:
feature: expose internal TrayIcon

petersamokhin committed:
feature: expose internal TrayIcon

FabianLars commented:
Thanks for the PR! Would you mind adding a comment similar to the one we have here? https://docs.rs/tauri/latest/tauri/webview/struct.WebviewWindow.html#method.with_webview If we expose the TrayIcon instance like this updating tray-icon would be a breaking change strictly speaking.
A comment about the main thread requirement when interacting with the instance would probably be nice to have too.

Also, I'd appreciate a changefile (check the .changes dir for examples) :)

petersamokhin committed:
feature: expose internal TrayIcon

petersamokhin commented:
@FabianLars definitely makes sense, thanks for noting! I added the changes document, and also updated the newly introduced fn to follow the `with_webview` approach that also explicitly handles the main thread.

There's one little difference from `with_webview`: the new fn returns whatever the passed `f` returns, instead of always returning `()`, just for convenience. Consumers still can return `()` if they wish but do it explicitly (even with a `;`). Let me know if you want this changed, not insisting here.

Thanks for the prompt response!

petersamokhin committed:
feature: expose internal TrayIcon

petersamokhin commented:
Fixed the build — somehow an import was removed from tests, reverted 🤷

Legend-Master commented on crates/tauri/src/tray/mod.rs:
```diff
@@ -603,6 +603,18 @@ impl<R: Runtime> TrayIcon<R> {
     })
   }
+
+  /// Get the inner tray icon.
```
```suggestion
  /// Do something with the inner [`tray_icon::TrayIcon`] on main thread
```

petersamokhin replied:
@Legend-Master fixed 🫡

Legend-Master commented on crates/tauri/src/tray/mod.rs:
```diff
@@ -603,6 +603,18 @@ impl<R: Runtime> TrayIcon<R> {
+  /// Therefore, it’s recommended to pin Tauri to at least a minor version when you’re using `with_inner_tray_icon`.
+  pub fn with_inner_tray_icon<F, T>(&self, f: F) -> crate::Result<T>
+  where
+    F: FnOnce(tray_icon::TrayIcon) -> T + Send + 'static,
```
I think we can avoid the clone here right? `&tray_icon::TrayIcon` should be able to do everything needed, and if people want, they could in rare cases clone it themselves

petersamokhin replied:
@Legend-Master fixed 🫡

petersamokhin committed:
feature: expose internal TrayIcon

Legend-Master commented on crates/tauri/src/tray/mod.rs:
```diff
@@ -623,7 +635,6 @@ mod tests {
   fn tray_event_json_serialization() {
     // NOTE: if this test is ever changed, you probably need to change `TrayIconEvent` in JS as well
 
-    use super::*;
```
😂

petersamokhin replied:
aaaaaaaaaaaaaaaaaa

petersamokhin replied:
sorry for that 💀  
idk jetbrains pls stop

petersamokhin replied:
fixed

petersamokhin committed:
feature: expose internal TrayIcon
